Source/Core/Core/PowerPC/BreakPoints.cpp

```diff
@@ -10,12 +10,57 @@
 #include <string>
 #include <vector>
 
+#include <expr.h>
+
 #include "Common/CommonTypes.h"
 #include "Common/DebugInterface.h"
 #include "Common/Logging/Log.h"
 #include "Core/Core.h"
 #include "Core/PowerPC/JitInterface.h"
 #include "Core/PowerPC/MMU.h"
+#include "Core/PowerPC/PowerPC.h"
+
+static bool ParseGPR(const char* name, int* index)
+{
+  if (name[0] == 'r' && isdigit(name[1]) &&
+      (name[2] == '\0' || isdigit(name[2]) && (name[3] == '\0')))
+  {
+    int idx = name[1] - '0';
+    if (name[2] != '\0')
+    {
+      idx = idx * 10 + (name[2] - '0');
+    }
+    if (idx < 32)
+    {
+      *index = idx;
+      return true;
+    }
+  }
+
+  *index = 0;
+  return false;
+}
+
+static bool EvaluateExpression(const std::string& expression)
+{
+  static expr_var_list vars = {0};
```

Comment:
Maybe organize this such that a local static isn't required. This inherently makes the function thread-unsafe.

Reply:
I don't intend to leave it this way. That said, code like this has been thread safe on conforming compilers since C++11, thanks to [magic statics](https://blog.mbedded.ninja/programming/languages/c-plus-plus/magic-statics/).

Reply:
That only applies for *initialization* of the local static. It does not mean data races do not matter in relation to the local static. If you call this in two different threads, you will clobber the data within the variable list (which is a race, and by extension, undefined behavior).

Reply:
Ah, you are correct, of course. I assumed your comment was about initialization since that has historically been an issue with statics in C++.

The scope and lifetime of expression variables is still up in the air. In the current implementation they all go in persistent global scope that is shared between all expressions. I believe they are only accessed from the CPU thread at the moment, but if they start to become exposed in various places in the UI, we'll definitely need a synchronization primitive to ensure thread safety.

Reply:
That said, I don't see any much of anything in the way of synchronization in this area, today. PowerPC::breakpoints is accessed freely from the UI thread while the game is running.... so that's not great. 😅

Reply:
Any reason this even needs to be static? The expression is destructed at the end of this function anyway...

Reply:
Slapping static on the variable list was a hack to create a global scope for expression variables. I believe this idea has utility, but it was not meant to be the final implementation. I should have left a comment in the code communicating this intent.